[RISCV] Optimize the spill/reload of segment registers

wangpc-pp opened the pull request:
The simplest way is:

1. Save `vtype` to a scalar register.
2. Insert a `vsetvli`.
3. Use segment load/store.
4. Restore `vtype` via `vsetvl`.

But `vsetvl` is usually slow, so this PR is not in this way.

Instead, we use wider whole load/store instructions if the register
encoding is aligned. We have done the same optimization for COPY in
https://github.com/llvm/llvm-project/pull/84455.

We found this suboptimal implementation when porting some video codec
kernels via RVV intrinsics.

wangpc-pp committed:
[RISCV] Optimize the spill/reload of segment registers

The simplest way is:

1. Save `vtype` to a scalar register.
2. Insert a `vsetvli`.
3. Use segment load/store.
4. Restore `vtype` via `vsetvl`.

But `vsetvl` is usually slow, so this PR is not in this way.

Instead, we use wider whole load/store instructions if the register
encoding is aligned. We have done the same optimization for COPY in
https://github.com/llvm/llvm-project/pull/84455.

We found this suboptimal implementation when porting some video codec
kernels via RVV intrinsics.

preames reviewed (commented):
Good idea, a few detail bits inline, but looks straight forward.

topperc commented on llvm/lib/Target/RISCV/RISCVRegisterInfo.cpp:
```diff
@@ -403,71 +419,65 @@ void RISCVRegisterInfo::lowerVSPILL(MachineBasicBlock::iterator II) const {
+        }
+        uint32_t ShiftAmount = Log2_32(PreSavedNum);
+        if (ShiftAmount == 0)
+          Step = VLENB;
```
We probably shouldn't have created a virtual register for this case.

topperc commented on llvm/lib/Target/RISCV/RISCVRegisterInfo.cpp:
```diff
@@ -486,65 +496,63 @@ void RISCVRegisterInfo::lowerVRELOAD(MachineBasicBlock::iterator II) const {
+  while (I != NumRegs) {
+    auto [LMulReloaded, RegClass, Opcode] =
+        getSpillReloadInfo(I, NumRegs, DestEncoding, false);
+    auto [NumReloaded, _] = RISCVVType::decodeVLMUL(LMulReloaded);
```
This looks very similar to the Spill code. Is there a common helper function we could extract from here?

wangpc-pp committed:
Add tests for 124/241 sequences

wangpc-pp committed:
First round of addressing comments

1. Add `NumRemaining`.
2. Rename `NewSize` to `VRegSize`.
3. Add argument comments.
4. Don't create `Step` for `ShiftAmount==0` case.

lukel97 commented on llvm/lib/Target/RISCV/RISCVRegisterInfo.cpp:
```diff
@@ -486,65 +498,66 @@ void RISCVRegisterInfo::lowerVRELOAD(MachineBasicBlock::iterator II) const {
+      Register Step;
+      if (auto VLEN = STI.getRealVLen()) {
+        Step = MRI.createVirtualRegister(&RISCV::GPRRegClass);
+        const int64_t VLENB = *VLEN / 8;
+        int64_t Offset = VLENB * PreReloadedNum;
```
Nit, inline this to avoid the name collision with `Register VLENB` above?

```suggestion
        int64_t Offset = (*VLEN / 8) * PreReloadedNum;
```

lukel97 commented on llvm/lib/Target/RISCV/RISCVRegisterInfo.cpp:
```diff
@@ -486,65 +498,66 @@ void RISCVRegisterInfo::lowerVRELOAD(MachineBasicBlock::iterator II) const {
+        .addMemOperand(MF.getMachineMemOperand(OldMMO, OldMMO->getOffset(),
+                                               VRegSize * NumReloaded));
+
+    PreReloadedNum = NumReloaded;
+    DestEncoding += NumReloaded;
+    I += NumReloaded;
```
Instead of keeping track of PreReloadedNum, is it possible to increment Base at the end of the loop if there's still more registers? Something like


```suggestion
    DestEncoding += NumReloaded;
    I += NumReloaded;
    if (I == NumRegs)
      break;
    // the code above in if (PreReloadedNum) { ...
    Base = NewBase;
  }
```

wangpc-pp replied:
My initial implementation was like this, I changed it to increase the base pointer at the beginning of loop because we need to know if there will be more registers to handle, so that we can set `IsKill` properly.

lukel97 reviewed (commented):
Makes sense to me

wangpc-pp committed:
Fold spill/reload implementations into one function

wangpc-pp committed:
Inline const VLENB

wangpc-pp committed:
Set IsKill for last step reg

lukel97 commented on llvm/lib/Target/RISCV/RISCVRegisterInfo.cpp:
```diff
@@ -403,148 +419,77 @@ void RISCVRegisterInfo::lowerVSPILL(MachineBasicBlock::iterator II) const {
+  unsigned PreHandledNum = 0;
+  unsigned I = 0;
+  while (I != NumRegs) {
+    auto [LMulHandled, RegClass, Opcode] =
+        getSpillReloadInfo(NumRegs - I, RegEncoding, IsSpill);
+    auto [RegNumHandled, _] = RISCVVType::decodeVLMUL(LMulHandled);
```
Do we need to return LMulHandled? Can we compute it from RegClass with `getLMul(RegClass.TSFlags)`

lukel97 replied:
Actually I don't think we need RegClass either, we can get it from `TII->getRegClass(TII->get(Opcode), 0, TRI, MF)` That way all you need to return in `getSpillReloadInfo` is the opcode.

wangpc-pp committed:
Update IsKill state

wangpc-pp committed:
Fix test

wangpc-pp committed:
Combine BuildMIs

lukel97 reviewed (approved):
LGTM

wangpc-pp committed:
Add IsLast
